src/loss.jl

```diff
@@ -0,0 +1,120 @@
+# Nonlinear shrinkage estimation can be described in terms of a loss function measuring the error between
+# the target and the sample eigenvalues. In:
+#    Donoho, D.L., Gavish, M. and Johnstone, I.M., 2018.
+#    Optimal shrinkage of eigenvalues in the spiked covariance model. Annals of statistics, 46(4), p.1742.
+# there is a systematic analysis of different loss functions; their classification scheme is encoded here.
+
+# Implementation note:
+# While we could parametrize all these different loss functions in the type system and use dispatch,
+# that would induce needless specialization: every function that took a `LossFunction` would have to be
+# specialized, even if the the argument encoding the loss function is merely "passed-through."
+# So instead, we hide details from the type system, dividing the Donoho classification scheme into just two types,
+# `NormLoss` and `StatLoss`, and use a fast runtime check to determine which shrinkage function to use.
+
+abstract type LossFunction end
+
+"""
+    NormLoss(norm::Symbol, pivotidx::Int)
+
+Specify a loss function for which the estimated covariance will be optimal. `norm` is one of
+`:L1`, `:L2`, or `:Linf`, and `pivotidx` is an integer from 1 to 7, as specified in Table 1 (p. 1755)
+of Donoho et al. (2018). In the table below, `A` and `B` are the target and sample covariances,
+respectively, and the loss function is the specified norm on the quantity in the `pivot` column:
+
+| `pivotidx` | `pivot` | Notes |
+|------------|---------|-------|
+| 1          | `A - B`     | |
+| 2          | `A⁻¹ - B⁻¹`     | |
+| 3          | `A⁻¹ B - I`     | Not available for `:L1` |
+| 4          | `B⁻¹ A - I`     | Not available for `:L1` |
+| 5          | `A⁻¹ B + B⁻¹ A - 2I` | Not supported |
+| 6          | `sqrt(A) \\ B / sqrt(A) - I` | |
+| 7          | `log(sqrt(A) \\ B / sqrt(A))` | Not supported |
+
+See also [`StatLoss`](@ref).
+
+Reference:
+    Donoho, D.L., Gavish, M. and Johnstone, I.M., 2018.
+    Optimal shrinkage of eigenvalues in the spiked covariance model. Annals of statistics, 46(4), p.1742.
+"""
+struct NormLoss <: LossFunction
```

Comment:
Please add these docstrings to https://github.com/mateuszbaran/CovarianceEstimation.jl/blob/master/docs/src/lib/public.md#public-interface